flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -908,38 +909,56 @@ public void reportCheckpointMetrics(
-                                                                + " instead of FINISHED."));
-                                    }
-                                    return jobstatus;
-                                });
-
         return savepointFuture
-                .thenCompose((path) -> terminationFuture.thenApply((jobStatus -> path)))
+                .thenCompose(
+                        path ->
+                                executionGraphTerminationFuture
+                                        .handle(
+                                                (executionStates, throwable) -> {
+                                                    Set<ExecutionState> nonFinishedStates =
+                                                            extractNonFinishedStates(
+                                                                    executionStates);
+                                                    if (throwable != null) {
+                                                        log.info(
+                                                                "Failed during stopping job {} with a savepoint. Reason: {}",
+                                                                jobGraph.getJobID(),
+                                                                throwable.getMessage());
+                                                        throw new CompletionException(throwable);
+                                                    } else if (!nonFinishedStates.isEmpty()) {
+                                                        log.info(
+                                                                "Failed while stopping job {} after successfully creating a savepoint. A global failover is going to be triggered. Reason: One or more states ended up in the following termination states instead of FINISHED: {}",
+                                                                jobGraph.getJobID(),
+                                                                nonFinishedStates);
+                                                        FlinkException
+                                                                inconsistentFinalStateException =
+                                                                        new FlinkException(
+                                                                                String.format(
+                                                                                        "Inconsistent execution state after stopping with savepoint. A global fail-over was triggered to recover the job %s.",
+                                                                                        jobGraph
+                                                                                                .getJobID()));
+                                                        executionGraph.failGlobal(
+                                                                inconsistentFinalStateException);
```

Comment:
I'm a bit concerned about thread-safety here: the statement seems non-thread-safe but it can be executed by any thread that terminates some `Execution`.

Previously, this block didn't access any state.

Should we use `handleAsync` with `mainThreadExecutor` here?

Reply:
Interesting finding. It was actually like that beforehand (using `handleAsync` in the `mainThreadExecutor`). But I ran into issues in `SavepointITCase.testStopSavepointWithBoundedInputConcurrently` which failed randomly. Switching to `handle` fixed it - hence, the change.

Reply:
The test was unstable on master (for some other reason) so I've just replaced it with a lower-level test.
So I think `handleAsync` shouldn't have this problem now.

Reply:
Cool, I rebased the branch and reverted back to `handleAsync`. One question, is there a reason to keep the leftover code from the IT test you reverted? If not, I would add a cleanup hotfix to the branch to remove things like `SavepointITCase.ischeckpointcoordinatorshutdownError` and the extensions you introduced for `InfiniteTestSource`. It looks like they we're only used for `SavepointITCase.testStopSavepointWithBoundedInputConcurrently`.

Reply:
Good point, I'll do that!